multicluster/controllers/multicluster/core/remote_cluster.go

```diff
@@ -0,0 +1,353 @@
+	Namespace string
+
+	// connectivity status, should it be the status?
+	connected bool
+
+	localClusterManager LocalClusterManager
+
+	remoteClusterManager RemoteClusterManager
+
+	stopFunc context.CancelFunc
+}
+
+/**
+ * A remote cluster is a leader in the ClusterSet Spec. This creates a remoteCluster instance which will
+ * use the secret and access credentials for the leader to connect to its common area.
+ */
+func NewRemoteCluster(clusterID common.ClusterID, clusterSetID common.ClusterSetID, url string, secretName string,
+	scheme *runtime.Scheme, localClusterManager LocalClusterManager, remoteClusterManager RemoteClusterManager,
+	clusterSetNamespace string, configNamespace string) (common.CommonArea, error) {
+	klog.Infof("Create remote cluster for", "cluster", clusterID)
+
+	// Secret associated with this local cluster must be copied from remote cluster to local cluster
+	// so that the secret data can be obtained.
+	// read/decode secret (get token and crt)
+	// TODO: what will be the namespace in the local cluster? use the same namespace where cluster set is configured
+	crtData, token, err := getSecretCACrtAndToken(configNamespace, secretName)
+	if err != nil {
+		return nil, err
+	}
+
+	// create manager for the member cluster
+	klog.Infof("Connecting", "cluster", clusterID, "url", url)
+	config, err := clientcmd.BuildConfigFromFlags(url, "")
+	if err != nil {
+		return nil, err
+	}
+	config.BearerToken = string(token)
+	config.CAData = crtData
+	mgr, err := ctrl.NewManager(config, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		Namespace:          clusterSetNamespace,
+	})
+	if err != nil {
+		klog.Error(err, "unable to create new manager", "cluster", clusterID)
+		return nil, err
+	}
+
+	remoteClient, e := client.New(config, client.Options{Scheme: scheme})
+	if e != nil {
+		return nil, e
+	}
+	cluster := &remoteCluster{
+		Client:               remoteClient,
+		ClusterManager:       mgr,
+		ClusterSetID:         clusterSetID,
+		ClusterID:            clusterID,
+		config:               config,
+		scheme:               scheme,
+		Namespace:            clusterSetNamespace,
+		connected:            false,
+		localClusterManager:  localClusterManager,
+		remoteClusterManager: remoteClusterManager,
+	}
+
+	remoteClusterManager.AddRemoteCluster(cluster)
+
+	return cluster, nil
+}
+
+/**
+ * When a member is added to a ClusterSet, a specific ServiceAccount is created on the
+ * leader cluster which allows the member to access into the common area. This ServiceAccount
+ * has an associated Secret which must be copied into the Member Cluster as an opaque secret.
+ * Name of this secret is part of the ClusterSet spec for this member. This method reads
+ * the Secret given by that name.
+ */
+func getSecretCACrtAndToken(namespace string, secretName string) ([]byte, []byte, error) {
+	var err error
+	mutex.Lock()
+	if secretClient == nil {
+		secretClient, err = client.New(ctrl.GetConfigOrDie(), client.Options{})
```

Comment:
This looks not-OO. Can we create client earlier, and pass it to RemoteCluster?

Reply:
Probably better to get the secret before NewRemoteCluster(). New func is typically for construction, and does not do block operations.

Reply:
Sure, I will change this.